src/Data/GADT/Internal.hs

```diff
@@ -189,6 +191,71 @@ instance (GRead a, GRead b) => GRead (a :+: b) where
                     | ("R1", s2) <- lex s1
                     , (r, t) <- greadsPrec 11 s2 ]) s
 
+-------------------------------------------------------------------------------
+-- HEq
+-------------------------------------------------------------------------------
+
+-- | Heterogenous lifted equality.
+--
+-- This class is stronger version of 'Eq1' from @base@
+--
+-- @
+-- class (forall a. Eq a => Eq (f a)) => Eq1 f where
+--     liftEq :: (a -> b -> Bool) -> f a -> f b -> Bool
+-- @
+--
+-- as we don't require a @a -> b -> Bool@ function.
+--
+-- Morally 'Eq1' should be a superclass of 'HEq', but it cannot be,
+-- as GHC wouldn't allow 'HEq' to be polykinded.
```

Comment:
I would cut this, it is not clear to clear what the argument to `liftEq` means in terms of this. e.g neither
```haskell
heq = liftEq (==)
```
or
```haskell
heq = liftEq \_ _ -> True
```
are satisfactory would-be laws.

Saying it is a `Eq1` where the argument is ignored is *also* not good, because
```haskell
data Foo :: Type -> Type where
  MkFoo :: [Int] -> Foo Int
  deriving Eq

instance Eq1 Foo where
  liftEq f (MkFoo x) (MkFoo y) = f x y

instance GEq Foo where
  geq (MkFoo x) (MkFoo y) = Refl <$ guard (x == y)
```
should be valid, but would be made non-complaint.

Reply:
BTW can we make a test suite with some of the weird examples we have been thinking of? That seems prudent.

Reply:
It means that lawful `Eq1` should be, `liftEq @f _ = heq @f`. Otherwise the `f` cannot have `HEq` instance. I'll spell that out explicitly.

Reply:
It's not type correct, but I guess you meant something like 

```haskell
data Foo :: Type -> Type where
  MkFoo :: Int -> Foo Int
```

Reply:
Complete example which type-checks with GHC-9.4

```haskell
{-# LANGUAGE GADTs, QuantifiedConstraints #-}

import Data.Kind

class (forall a. Eq a => Eq (f a)) => Eq1 f where
    liftEq :: (a -> b -> Bool) -> f a -> f b -> Bool

instance Eq1 [] where
    liftEq eq [] [] = True
    liftEq eq (x:xs) (y:ys) = eq x y && liftEq eq xs ys
    liftEq eq (_:_) [] = False
    liftEq eq [] (_:_) = False

data Foo :: Type -> Type where
    MkFoo :: [Int] -> Foo Int

deriving instance Eq (Foo a)

-- instance Eq1 Foo where
--   liftEq f (MkFoo x) (MkFoo y) = liftEq f x y

instance Eq1 Foo where
  liftEq _ (MkFoo x) (MkFoo y) = x == y
```

I think that the below `liftEq` instance is better. It's possible to write `liftEq f (MkFoo x) (MkFoo y) = liftEq f x y` instance, but it feels wrong.

Reply:
@phadej "feels wrong" is quite opinionated for a new revsion of an existing library. Remember, there has been no laws relating `Eq1` and `GEq` at all so far, and adding one so would be a breaking change.

Reply:
I personally find ignoring the function and using `(==)` instead very surprising. It seems like `EQ1` has something to do with `Align` and `Foldable` (and I think you have mentioned similar intuitions on this). Though I grant since `Foo` is not a functor, I still feel like one should use the `f` where possible as if it was.

I hope to take `EQ1` and friend out of `base` (after my last changes `Data.Functor.Classes` is easy turn into a leaf module). That can remove the urgency to relate these classes to them.